langkit/templates/pkg_implementation_body_ada.mako

```diff
@@ -2795,20 +2775,17 @@ package body ${ada_lib_name}.Implementation is
          end Get_Parent_Env;
 
       begin
-         % if sorted_nodes_adding_env:
-            % if not all_nodes_add_env:
-               if Node.Kind in ${' | '.join(sorted_nodes_adding_env)} then
+         <%self:case_dispatch
+           pred="${lambda n: n.inherited_env_spec and not n.abstract}">
+         <%def name="action(node)">
+            % if node.inherited_env_spec.adds_env:
                   return Get_Parent_Env;
-               else
-                  return Node.Self_Env;
-               end if;
             % else:
-               return Get_Parent_Env;
+                  return Node.Self_Env;
             % endif
-
-         % else:
-            return Node.Self_Env;
-         % endif
+         </%def>
+         <%def name="default()"> return Node.Self_Env; </%def>
+         </%self:case_dispatch>
```

Comment:
This will generate one `when` clause for each node that has an effective env spec. So lots of redundant code. What about instead computing the list of kinds for which this must return `Get_Parent_Env` and generate at most two clauses instead? (like the previous version did)

Reply:
Well the generated code might be redundant, but the template is made much simpler, and the resulting machine code will very probably be just as simple. Given the number of time we have gotten this function wrong, I'd rather keep this simple and readable (like the previous version wasn't IMHO).

Reply:
But that’s unrelated: the previous version was wrong because its author (me, IIRC) did not reason correctly about what it should do, yielding a complex and wrong algorithm in the process. I think generating a simple `if` check instead of a big `case` block won’t make the above more complex.

Reply:
Well it implies re-encoding some of the logic about iterating on nodes and collapsing branches that is not trivial, and is automated by the `case..` block. But probably not undoable indeed.